File: backend/app/schemas/meal.py
from datetime import datetime
from decimal import Decimal
from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field, field_serializer, field_validator

from app.models.enums import MealType, NutrientName
from app.utils.pagination import PaginatedResponse
from app.utils.validators import ensure_utc

DecFloat = Annotated[Decimal, Field(ge=0, max_digits=10, decimal_places=2)]
MicroAmount = Annotated[Decimal, Field(ge=0, max_digits=12, decimal_places=4)]


class DecimalJsonModel(BaseModel):
    @field_serializer("*", when_used="json")
    def serialize_decimals(self, value: object) -> object:
        return float(value) if isinstance(value, Decimal) else value


class MicronutrientCreate(BaseModel):
    nutrient_name: NutrientName
    amount: MicroAmount
    unit: str = Field(min_length=1, max_length=20)


class MicronutrientPublic(DecimalJsonModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    nutrient_name: str
    amount: Decimal
    unit: str


class FoodEntryCreate(BaseModel):
    food_name: str = Field(min_length=1, max_length=255)
    quantity: DecFloat
    unit: str = Field(min_length=1, max_length=40)
    calories: DecFloat
    protein: DecFloat
    carbohydrates: DecFloat
    fat: DecFloat
    fiber: DecFloat = Decimal("0")
    sugar: DecFloat = Decimal("0")
    micronutrients: list[MicronutrientCreate] = Field(default_factory=list)


class FoodEntryPublic(DecimalJsonModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    food_name: str
    quantity: Decimal
    unit: str
    calories: Decimal
    protein: Decimal
    carbohydrates: Decimal
    fat: Decimal
    fiber: Decimal
    sugar: Decimal
    micronutrients: list[MicronutrientPublic]


class MealCreate(BaseModel):
    meal_type: MealType
    consumed_at: datetime
    notes: str | None = Field(default=None, max_length=2000)
    food_entries: list[FoodEntryCreate] = Field(min_length=1)

    @field_validator("consumed_at")
    @classmethod
    def timezone_aware(cls, value: datetime) -> datetime:
        return ensure_utc(value)

    @field_validator("notes")
    @classmethod
    def empty_notes(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        return stripped or None


class MealUpdate(MealCreate):
    """Full replacement of a meal and its food entries."""


class MealTotals(DecimalJsonModel):
    calories: Decimal
    protein: Decimal
    carbohydrates: Decimal
    fat: Decimal
    fiber: Decimal
    sugar: Decimal


class MealPublic(DecimalJsonModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    meal_type: MealType
    consumed_at: datetime
    notes: str | None
    food_entries: list[FoodEntryPublic]
    totals: MealTotals
    created_at: datetime
    updated_at: datetime


MealListResponse = PaginatedResponse[MealPublic]

File: frontend/src/types/meal.ts
export const MEAL_TYPES = ["BREAKFAST", "LUNCH", "DINNER", "SNACK"] as const;
export type MealType = (typeof MEAL_TYPES)[number];

export const NUTRIENT_NAMES = [
  "Vitamin A",
  "Vitamin B1",
  "Vitamin B2",
  "Vitamin B3",
  "Vitamin B6",
  "Vitamin B12",
  "Vitamin C",
  "Vitamin D",
  "Vitamin E",
  "Vitamin K",
  "Calcium",
  "Iron",
  "Magnesium",
  "Potassium",
  "Zinc",
  "Sodium",
] as const;

export type NutrientName = (typeof NUTRIENT_NAMES)[number];

export type Micronutrient = {
  id?: number;
  nutrient_name: string;
  amount: number;
  unit: string;
};

export type FoodEntry = {
  id?: number;
  food_name: string;
  quantity: number;
  unit: string;
  calories: number;
  protein: number;
  carbohydrates: number;
  fat: number;
  fiber: number;
  sugar: number;
  micronutrients: Micronutrient[];
};

export type MealTotals = {
  calories: number;
  protein: number;
  carbohydrates: number;
  fat: number;
  fiber: number;
  sugar: number;
};

export type Meal = {
  id: number;
  user_id: number;
  meal_type: MealType;
  consumed_at: string;
  notes: string | null;
  food_entries: FoodEntry[];
  totals: MealTotals;
  created_at: string;
  updated_at: string;
};

export type MealPayload = {
  meal_type: MealType;
  consumed_at: string;
  notes: string | null;
  food_entries: Array<Omit<FoodEntry, "id">>;
};

export type MealListParams = {
  page?: number;
  page_size?: number;
  date?: string;
  start_date?: string;
  end_date?: string;
  meal_type?: MealType | "";
  q?: string;
};
